primitives/src/evm.rs

```diff
@@ -222,6 +230,58 @@ impl TryFrom<CurrencyId> for EvmAddress {
 	}
 }
 
+pub fn decode_gas_price(gas_price: u64, gas_limit: u64, tx_fee_per_gas: u128) -> Option<(u128, u32)> {
+	// ensure gas_price >= 100 Gwei
+	if u128::from(gas_price) < tx_fee_per_gas {
+		return None;
+	}
+
+	let mut tip: u128 = 0;
+	let mut actual_gas_price = gas_price;
+	const ONE_GWEI: u64 = 1_000_000_000u64;
+
+	// percentage
+	let tip_number = gas_price.checked_div(ONE_GWEI)?.checked_sub(100)?;
+	if !tip_number.is_zero() {
+		actual_gas_price = gas_price.checked_sub(tip_number.checked_mul(ONE_GWEI)?)?;
+		tip = actual_gas_price
+			.checked_mul(gas_limit)?
+			.checked_mul(tip_number)?
+			.checked_div(100)? // percentage
+			.checked_div(1_000_000)? // ACA decimail is 12, ETH decimail is 18
+			.into();
+	}
+
+	// valid_until max is 999_999_999
+	let valid_until: u32 = Into::<u128>::into(actual_gas_price)
+		.checked_sub(tx_fee_per_gas)?
+		.try_into()
+		.ok()?;
```

Comment:
`gasPrice = 100 Gwei = 100_000_000_000`,  so valid_until max is `999_999_999`. 
`BlockNumber = u32` , `u32::MAX; // 4_294_967_295u32`.  Decimal is not enough here, but it should be fine.

Reply:
with 6s block time, the world is going to end in 190 years? 💀

Reply:
:joy:

Reply:
Now use 10 Gwei(`10_000_000_000`) to store valid_until, and `valid_until = min(u32::MAX, valid_until)`

Reply:
@xlc please take a look